st2common/st2common/metrics/metrics.py

```diff
@@ -0,0 +1,162 @@
+# Licensed to the StackStorm, Inc ('StackStorm') under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import absolute_import
+from functools import wraps
+
+from oslo_config import cfg
+from oslo_config.cfg import NoSuchOptError
+from stevedore.exception import NoMatches, MultipleMatches
+
+from st2common.constants.metrics import METRICS_COUNTER_SUFFIX, METRICS_TIMER_SUFFIX
+from st2common.util.loader import get_plugin_instance
+from st2common.util.date import get_datetime_utc_now
+
+
+PLUGIN_NAMESPACE = 'st2common.metrics.driver'
+
+
+class BaseMetricsDriver(object):
+    """ Base class for driver implementations for metric collection
+    """
+    def time(self, key, time):
+        """ Timer metric
+        """
+        pass
+
+    def inc_counter(self, key, amount=1):
+        """ Increment counter
+        """
+        pass
+
+    def dec_counter(self, key, amount=1):
+        """ Decrement metric
+        """
+        pass
+
+
+class Timer(object):
+    """ Timer context manager for easily sending timer statistics.
+    """
+    def __init__(self, key):
+        assert isinstance(key, str)
+        assert len(key) > 0
+        self.key = key
+        self._metrics = METRICS
+        self._start_time = None
+
+    def send_time(self, key=None):
+        """ Send current time from start time.
+        """
+        time_delta = get_datetime_utc_now() - self._start_time
+
+        if key:
+            assert isinstance(key, str)
+            self._metrics.time(key, time_delta.total_seconds())
```

Comment:
(just thinking out loud)

I assume that in the context where we will be using that seconds are enough and there is no need for micro / nano seconds, although miliseconds might not be a bad idea.